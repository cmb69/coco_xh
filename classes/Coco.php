<?php

/**
 * Copyright 2012-2023 Christoph M. Becker
 *
 * This file is part of Coco_XH.
 *
 * Coco_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Coco_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Coco_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Coco;

use Coco\Infra\CsrfProtector;
use Coco\Infra\Repository;
use Coco\Infra\Request;
use Coco\Infra\XhStuff;
use Coco\Infra\View;
use Coco\Logic\Util;
use Coco\Value\Html;
use Coco\Value\Response;

class Coco
{
    /** @var Repository */
    private $repository;

    /** @var CsrfProtector */
    private $csrfProtector;

    /** @var XhStuff */
    private $xhStuff;

    /** @var View */
    private $view;

    public function __construct(
        Repository $repository,
        CsrfProtector $csrfProtector,
        XhStuff $xhStuff,
        View $view
    ) {
        $this->repository = $repository;
        $this->csrfProtector = $csrfProtector;
        $this->xhStuff = $xhStuff;
        $this->view = $view;
    }

    public function __invoke(Request $request, string $name, string $config, string $height): Response
    {
        if (!Util::isValidCocoName($name)) {
            return Response::create($this->view->message("fail", "error_invalid_name") . "\n");
        }
        if ($request->s() < 0) {
            return Response::create("");
        }
        switch ($request->cocoAction($name)) {
            default:
                return $this->show($request, $name);
            case "edit":
                return $this->edit($request, $name, $config, $height);
            case "do_edit":
                return $this->update($request, $name, $config, $height);
        }
    }

    private function show(Request $request, string $name): Response
    {
        $content = $this->repository->find($name, $request->s());
        $content = $this->xhStuff->evaluateScripting($content);
        $search = $request->search();
        if ($search !== "") {
            $words = Util::parseSearchTerm($search);
            $content = $this->xhStuff->highlightSearchWords($words, $content);
        }
        return Response::create($content);
    }

    private function edit(Request $request, string $name, string $config, string $height): Response
    {
        $content = $this->repository->find($name, $request->s());
        return Response::create($this->renderEditor($name, $config, $height, $content));
    }

    private function update(Request $request, string $name, string $config, string $height): Response
    {
        $this->csrfProtector->check();
        $text = $request->cocoText($name);
        if ($this->repository->save($name, $request->s(), $text)) {
            return Response::redirect($request->url()->absolute());
        }
        return Response::create(
            $this->view->message("fail", "error_save", $this->repository->filename($name))
            . $this->renderEditor($name, $config, $height, $text)
        );
    }

    private function renderEditor(string $name, string $config, string $height, string $content): string
    {
        $id = "coco_text_$name";
        $editor = $this->xhStuff->replaceEditor($id, $config);
        return $this->view->render("edit_form", [
            "id" => $id,
            "name" => $name,
            "style" => "width:100%; height:$height",
            "content" => $content,
            "editor" => $editor !== false ? Html::of($editor) : false,
            "csrf_token" => $this->csrfProtector->token(),
        ]);
    }
}
