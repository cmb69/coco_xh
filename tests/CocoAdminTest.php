<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Coco_XH.
 *
 * Coco_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Coco_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Coco_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Coco;

use ApprovalTests\Approvals;
use Coco\Infra\CsrfProtector;
use Coco\Infra\Repository;
use Coco\Infra\Request;
use Coco\Infra\View;
use PHPUnit\Framework\TestCase;

class CocoAdminTest extends TestCase
{
    public function testRendersCocoOverview(): void
    {
        $sut = new CocoAdmin($this->repository(), $this->csrfProtector(), $this->view());
        $response = $sut($this->request());
        $this->assertEquals("Coco – Co-Contents", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testRendersDeleteConfirmation(): void
    {
        $sut = new CocoAdmin($this->repository(), $this->csrfProtector(), $this->view());
        $request = $this->request();
        $request->method("cocoAdminAction")->willReturn("delete");
        $request->method("cocoNames")->willReturn(["foo"]);
        $response = $sut($request);
        $this->assertEquals("Coco – Co-Contents", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testSuccessfulDeletionRedirects(): void
    {
        $sut = new CocoAdmin($this->repository(), $this->csrfProtector(true), $this->view());
        $request = $this->request();
        $request->method("cocoAdminAction")->willReturn("do_delete");
        $request->method("cocoNames")->willReturn(["foo"]);
        $response = $sut($request);
        $this->assertEquals("http://example.com/?coco&admin=plugin_main", $response->location());
    }

    public function testFailureToDeleteIsReported(): void
    {
        $repository = $this->repository(false);
        $sut = new CocoAdmin($repository, $this->csrfProtector(true), $this->view());
        $request = $this->request();
        $request->method("cocoAdminAction")->willReturn("do_delete");
        $request->method("cocoNames")->willReturn(["foo"]);
        $response = $sut($request);
        $this->assertEquals("Coco – Co-Contents", $response->title());
        $this->assertStringContainsString("./content/coco/foo.htm could not be deleted!", $response->output());
    }

    private function request(): Request
    {
        $request = $this->createMock(Request::class);
        $request->method("sn")->willReturn("/");
        return $request;
    }

    private function repository(bool $deleted = true): Repository
    {
        $repository = $this->createMock(Repository::class);
        $repository->method("findAllNames")->willReturn(["foo", "bar"]);
        $repository->method("findAllBackups")->willReturn([["foo", "20230306_120000"]]);
        $repository->method("delete")->willReturn($deleted);
        $repository->method("filename")->willReturnOnConsecutiveCalls(
            "./content/coco/20230306_120000_foo.htm",
            "./content/coco/foo.htm"
        );
        return $repository;
    }

    private function csrfProtector(bool $check = false): CsrfProtector
    {
        $csrfProtector = $this->createMock(CsrfProtector::class);
        $csrfProtector->method("token")->willReturn("eee5e668b3bcc9b71a9e4cc1aa76393f");
        $csrfProtector->expects($check ? $this->once() : $this->never())->method("check");
        return $csrfProtector;
    }

    private function view(): View
    {
        return new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["coco"]);
    }
}
