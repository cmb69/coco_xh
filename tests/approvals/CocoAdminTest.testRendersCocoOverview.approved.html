<!-- coco administration -->
<h1>Coco â€“ Co-Contents</h1>
<form id="coco_admin_cocos" action="/" method="get">
  <input type="hidden" name="selected" value="coco"/>
  <input type="hidden" name="admin" value="plugin_main"/>
  <table>
    <tr>
      <th>Coco</th>
    </tr>
    <tr>
      <td>
        <label>
          <input type="checkbox" name="coco_name[]" value="foo">
          <span>foo</span>
        </label>
      </td>
    </tr>
    <tr>
      <td>
        <label>
          <input type="checkbox" name="coco_name[]" value="bar">
          <span>bar</span>
        </label>
      </td>
    </tr>
  </table>
  <p><button name="action" value="delete">Delete</button></p>
</form>
